test/fortran/f_c_string/Test_f_c_string_t_mod.f90

```diff
@@ -0,0 +1,41 @@
+module Test_f_c_string_t_mod
+    use fortran_test_framework_mod, only : assertEqual, assert
+    use f_c_string_t_mod, only : f_c_string_t, strlen
+    use Test_f_c_string_i_mod, only : to_upper
+    use iso_c_binding, only : c_null_char, c_ptr
+    implicit none
+    contains
+    subroutine Test_f_c_string_t()
+        type(f_c_string_t), allocatable :: f_c_string1
+        type(f_c_string_t), allocatable :: f_c_string2
+        type(f_c_string_t), allocatable :: f_c_string_empty
+        character(len=:), allocatable :: f_string
+        character(len=:), allocatable :: f_string_upper
+        character(len=:), allocatable :: f_string_ref
+        character(len=:), allocatable :: f_string_upper_ref
+        type(c_ptr) :: c_string
+        integer :: c_string_len, f_string_len, status
+
+        status = 0
+        f_string_ref = "foo"
+        f_string_upper_ref = "FOO"
+        f_string_len = len(f_string_ref)
+
+        allocate(f_c_string1)
+        allocate(f_c_string2)
+        allocate(f_c_string_empty)
+
+        c_string = f_c_string_empty%to_c("")
+        c_string_len= strlen(c_string)
+        call assertEqual(0, c_string_len, status, assert)
+
+        c_string = f_c_string1%to_c(f_string_ref)
+        c_string_len = strlen(c_string)
+        call assertEqual(f_string_len, c_string_len, status, assert)
```

Comment:
In my opinion, the `assertEqual` code is a needless complication compared with simply comparing values directly inline. For example: does `assertEqual` overwrite the value of `status` based on the success of this comparison, or does it only set `status` upon failure? If the former, then would an earlier failure (i.e., non-zero value of `status`) get lost if this current assertion succeeds? Having to search through the testing framework code to figure this out seems like it could be avoided by just comparing the strings inline here.

Reply:
The assertEqual function reduces redundancy by consolidating result verification, logging, and error handling for test scenarios. When it uses the assert subroutine as its handler, assert terminates the test via exit, which registers the failure with CTest.

However, for memcheck tests, we need a different assert handler that doesn't call exit on failure. Exiting the program during these tests causes memory leaks to appear as false positives, leading to misleading results.